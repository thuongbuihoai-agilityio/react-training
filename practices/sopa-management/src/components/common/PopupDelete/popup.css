.popup {
  position: absolute;
  background-color: var(--light);
  border-radius: var(--rounded-primary);
  padding: var(--px-extra-normal);
  top: 40%;
  right: var(--w-extra-large);
  width: 245px;
}

.popup-btn {
  text-align: center;
  margin-top: var(--m-large);
}

.popup-cancel,
.popup-delete {
  padding: 10px 20px;
  border: none;
  border-radius:  var(--rounded-tertiary);
}

.popup-cancel {
  margin-right: var(--m-extra-small);
}

.popup-delete {
  background-color: var(--text-primary);
}
