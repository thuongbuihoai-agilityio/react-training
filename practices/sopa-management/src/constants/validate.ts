const CONFIRM_MESSAGE = {
  ADD_SUCCESS: 'Add product to cart successfully',
  UPDATE_SUCCESS: 'Update product in cart successfully',
  DELETE_SUCCESS: 'Delete product in cart successfully',
  CONFIRM_DELETE: 'Are you sure you want to delete?',
  LOGIN_SUCCESS: 'Login successfully'
}

const ERROR_MESSAGES = {
  REQUIRED: 'This field is required',
  EMAIL_FORMAT: 'Wrong email format',
  EMAIL_NOT_EXIST: 'Email does not exist',
  PASSWORD_FORMAT: 'Password must be more than 6 characters',
  PASSWORD_INCORRECT: 'Password is incorrect',
  ERROR_BOUNDARY: 'Something went wrong',
}

export {
  CONFIRM_MESSAGE,
  ERROR_MESSAGES,
};
