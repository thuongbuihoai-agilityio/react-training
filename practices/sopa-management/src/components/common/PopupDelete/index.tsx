import { memo } from 'react';

// Components
import {
  Button,
  Text
} from '@components/common';

// Styles
import './popup.css';

interface PopupDeleteProps {
  title?: string;
  onCancel: () => void;
  onDelete: () => void;
}

const PopupDelete = ({
  title = '',
  onCancel,
  onDelete,
}: PopupDeleteProps) => (
  <div data-testid='popup-delete' className='popup'>
    <Text text={title} />
    <div className='popup-btn'>
      <Button
        ariaLabel='Cancel'
        children='Cancel'
        className='popup-cancel'
        onClick={onCancel}
      />
      <Button
        ariaLabel='Delete'
        children='Delete'
        className='popup-delete'
        onClick={onDelete}
      />
    </div>
  </div>
);

export default memo(PopupDelete);
