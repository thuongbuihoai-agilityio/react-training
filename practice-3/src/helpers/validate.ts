import { ERROR_MSG } from "@/constants/message";
import { RULES } from "@/constants/rules";
import { FormProps } from "@/types/form";

const validateValue = (value: string, rule: string, errors: {[fieldName: string]: any}, fieldName: string) => {
  if (rule === RULES.REQUIRED) {
    if (value) {
      errors[fieldName].error = "";
      return;
    } else {
      errors[fieldName].error = ERROR_MSG.REQUIRED;
      return;
    }
  }

  if (rule === RULES.NUMBER) {
    if (typeof +value == RULES.NUMBER) {
      errors[fieldName].error += "";
    } else {
      errors[fieldName].error += ERROR_MSG.NUMBER;
      return;
    }
  }
}

const validate = (values: FormProps) => {
  const errors = {...values};
  (Object.keys(errors) as (keyof FormProps)[]).map(fieldName => {
    errors[fieldName].rules.map((rule: string)=>{
        validateValue(errors[fieldName].value, rule, errors, fieldName)
      });
    });

  return errors;
}

export {validate};