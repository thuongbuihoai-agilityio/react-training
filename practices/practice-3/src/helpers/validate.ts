import { ERROR_MSG } from "@/constants/message";
import { RULES } from "@/constants/rules";
import { FormProps } from "@/types/form";

const validateValue = (value: string, rule: string, errors: {[fieldName: string]: any}, fieldName: string) => {
  // check required
  if (rule === RULES.REQUIRED) {
    if (value) {
      errors[fieldName].error = "";
    } else {
      errors[fieldName].error = ERROR_MSG.REQUIRED;
    }
  }

  // check value is number
  if (rule === RULES.NUMBER) {
    if (typeof +value == RULES.NUMBER) {
      errors[fieldName].error += "";
    } else {
      errors[fieldName].error += ERROR_MSG.NUMBER;
    }
  }

  // check value is negative
  if(rule === RULES.NEGATIVE) {
    if(+value < 0) {
      errors[fieldName].error += ERROR_MSG.NUMBER;
    } else {
      errors[fieldName].error += "";
    }
  }
}

// handle validate
const validate = (values: FormProps) => {
  const errors = {...values};
  (Object.keys(errors) as (keyof FormProps)[]).map(fieldName => {
    errors[fieldName].rules.map((rule: string)=>{
        validateValue(errors[fieldName].value, rule, errors, fieldName)
      });
    });

  return errors;
}

export {validate};