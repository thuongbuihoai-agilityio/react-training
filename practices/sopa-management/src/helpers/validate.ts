import { REGEX } from "@constants/regex"
import { ERROR_MESSAGES } from "@constants/validate"

// Validate
export const VALIDATE = {
  email: {
    required: ERROR_MESSAGES.REQUIRED,
    pattern: {
      value: REGEX.EMAIL,
      message: ERROR_MESSAGES.EMAIL_FORMAT,
    },
  },
  password: {
    required: ERROR_MESSAGES.REQUIRED,
    pattern: {
      value: REGEX.PASSWORD,
      message: ERROR_MESSAGES.PASSWORD_FORMAT,
    },
  },
}
