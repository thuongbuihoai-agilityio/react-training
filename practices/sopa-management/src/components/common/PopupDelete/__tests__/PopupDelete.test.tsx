// Libs
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

// Components
import PopupDelete from '@common/PopupDelete';

// Constants
import { CONFIRM_MESSAGE } from '@constants/validate';

describe('PopupDelete component', () => {
  const props = {
    onCancel: jest.fn(),
    onDelete: jest.fn()
  };
  test('should render PopupDelete component', () => {
    const { getByTestId } = render(<PopupDelete {...props} />);
    expect(getByTestId('popup-delete')).toBeInTheDocument();
  });

  test('should render PopupDelete component', () => {
    const { getByTestId } = render(<PopupDelete title={CONFIRM_MESSAGE.CONFIRM_DELETE} {...props} />);
    expect(getByTestId('popup-delete')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<PopupDelete {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
