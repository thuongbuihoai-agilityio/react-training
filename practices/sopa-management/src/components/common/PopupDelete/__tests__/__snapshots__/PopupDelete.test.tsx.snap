// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupDelete component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="popup"
    data-testid="popup-delete"
  >
    <p
      class=" text-"
      data-testid="text"
    />
    <div
      class="popup-btn"
    >
      <button
        aria-label="Cancel"
        class="popup-cancel btn btn-"
        data-testid="button"
      >
        Cancel
      </button>
      <button
        aria-label="Delete"
        class="popup-delete btn btn-"
        data-testid="button"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
